src/active-response/firewalls/default-firewall-drop.c

```diff
@@ -0,0 +1,464 @@
+            command[COMMANDSIZE -1] = '\0';
+            snprintf(command, COMMANDSIZE - 1, "eval %s -v 4 -d", mkfilt_path);
+            res = system(command);
+            command[COMMANDSIZE -1] = '\0';
+            snprintf(command, COMMANDSIZE - 1, "eval %s -v 4 -u", mkfilt_path);
+            res = system(command);
+        }
+
+    } else {
+        write_debug_file(filename, "Invalid system");
+
+        free_vars();
+        return OS_INVALID;
+    }
+    write_debug_file(filename, "Ended");
+    free_vars();
+
+    return 0;
+}
+
+static void free_vars (){
+    cJSON_Delete(input_json);
+    os_free(srcip);
+    os_free(action);
+    os_free(iptables);
+}
+
+static void lock () {
+    int i=0;
+    int max_iteration = 50;
+    bool flag = true;
+    int read;
+
+    // Providing a lock.
+    while (flag){
+        char lock_pid_path[PATH_MAX];
+        FILE *pid_file;
+        pid_t current_pid;
+        snprintf(lock_path, PATH_MAX - 1, "%s%s", DEFAULTDIR, LOCK_PATH);
+        snprintf(lock_pid_path, PATH_MAX - 1, "%s%s", DEFAULTDIR, LOCK_FILE);
+
+        if (mkdir(lock_path, S_IRWXG) == 0) {
+            // Lock acquired (setting the pid)
+            pid_t pid = getpid();
+            pid_file = fopen(lock_pid_path, "w");
+            fprintf(pid_file, "%u", pid);
+            fclose(pid_file);
+            return;
+        }
+
+        // Getting currently/saved PID locking the file
+        if (pid_file = fopen(lock_pid_path, "r"), !pid_file) {
+            write_debug_file("default-firewall-drop", "Can not read pid file");
+            continue;
+        } else {
+            read = fscanf(pid_file, "%u", &current_pid);
+            fclose(pid_file);
+
+            if (read == 1) {
+                if (saved_pid == -1) {
+                    saved_pid = current_pid;
+                }
+
+                if (current_pid == saved_pid) {
+                    i++;
+                }
+
+            } else {
+                write_debug_file("default-firewall-drop", "Can not read pid file");
+                continue;
+            }
+        }
+
+        sleep(i);
+
+        i++;
+
+        // So i increments 2 by 2 if the pid does not change.
+        // If the pid keeps changing, we will increments one
+        // by one and fail after MAX_ITERACTION
+        if (i >= max_iteration) {
+            bool kill = false;
+            char output_buf[BUFFERSIZE];
+            char command[COMMANDSIZE];
+            snprintf(command, COMMANDSIZE -1, "pgrep -f %s", filename);
+            FILE *fp = popen(command, "r");
+            if (fp) {
+                while (fgets(output_buf, BUFFERSIZE, fp) != NULL) {
+                    pid_t pid = (pid_t)strtol(output_buf, NULL, 10);
+                    if (pid == current_pid) {
+                        memset(command, 0, COMMANDSIZE);
+                        snprintf(command, COMMANDSIZE -1, "kill -9 %u", pid);
+                        if (system(command) == 0) {
```

Comment:
Use `execv`.

Reply:
I also wouldn't check the result of the execution, the old script doesn't to that.